classes/source/source_url.php

```diff
@@ -0,0 +1,209 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * URL source.
+ *
+ * @package    tool_etl
+ * @copyright  2019 John Yao <johnyao@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+
+namespace tool_etl\source;
+
+use curl;
+use tool_etl\config_field;
+use tool_etl\data;
+use tool_etl\logger;
+
+
+defined('MOODLE_INTERNAL') || die;
+
+global $CFG;
+require_once($CFG->libdir . '/filelib.php');
+
+class source_url extends source_base {
+
+    /**
+     * Name of the source.
+     *
+     * @var string
+     */
+    protected $name = "URL";
+
+    /**
+     * Settings.
+     *
+     * @var array
+     */
+    protected $settings = array(
+        'address' => '',
+    );
+
+    /**
+     * A temp folder to save files.
+     *
+     * @var string
+     */
+    protected $filedir;
+
+    /**
+     * Date and time now.
+     *
+     * @var string
+     */
+    protected $now;
+
+    public function __construct(array $settings = array()) {
+        global $CFG;
+
+        parent::__construct($settings);
+
+        $this->now = date('YmdHis', time());
+
+        $this->filedir = $CFG->dataroot . DIRECTORY_SEPARATOR . $this->get_short_name();
+        check_dir_exists($this->filedir);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function extract() {
+        if ($this->is_available()) {
+            $this->data = new data([$this->get_file()]);
+        } else {
+            throw new \Exception($this->name . ' source is not available!');
+        }
+
+        return $this->data;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function is_available() {
```

Comment:
If this is just a test, I would not actually do the file transfer, just test that the url works.

If possibly I'd try and do a HEAD request instead of a GET so that if it is a massive file it is faster